modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+          yield elem;
+        }
+      }
+    }
+  }
+
+  iter these(param tag : iterKind) where tag == iterKind.leader {
+    coforall loc in targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      coforall segmentIdx in 0 .. #here.maxTaskPar {
+        ref segment = instance.bag.segments[segmentIdx];
+
+        if segment.acquireIfNonEmpty(STATUS_LOOKUP) {
+          // Create a snapshot...
+          var block = segment.headBlock;
+          var bufferSz = segment.nElems.read() : int;
+          var buffer = c_malloc(eltType, bufferSz);
+          var bufferOffset = 0;
+
+          while block != nil {
+            if bufferOffset + block.size > bufferSz {
+              halt("DistributedBag Internal Error: Snapshot attempt with bufferSz(", bufferSz, ") with offset bufferOffset(", bufferOffset + block.size, ")");
+            }
+            __primitive("chpl_comm_array_put", block.elems[0], here.id, buffer[bufferOffset], block.size);
+            bufferOffset += block.size;
+            block = block.next;
+          }
+
+          // Yield this chunk to be process...
+          segment.releaseStatus();
+          yield (bufferSz, buffer);
+          c_free(buffer);
+        }
+      }
+    }
+  }
+
+  iter these(param tag : iterKind, followThis) where tag == iterKind.follower {
+    var (bufferSz, buffer) = followThis;
+    for i in 0 .. #bufferSz {
+      yield buffer[i];
+    }
+  }
+}
+
+/*
+  A segment block is an unrolled linked list node that holds a contiguous buffer
+  of memory. Each segment block size *should* be a power of two, as we increase the
+  size of each subsequent unroll block by twice the size. This is so that stealing
+  work is faster in that majority of elements are confined to one area.
+*/
+pragma "no doc"
+class BagSegmentBlock {
+  type eltType;
+
+  // Contiguous memory containing all elements
+  var elems :  c_ptr(eltType);
+  var next : BagSegmentBlock(eltType);
+
+  // The capacity of this block.
+  var cap : int;
+  // The number of occupied elements in this block.
+  var size : int;
+
+  inline proc isEmpty {
+    return size == 0;
+  }
+
+  inline proc isFull {
+    return size == cap;
+  }
+
+  inline proc push(elt : eltType) {
+    if elems == nil {
+      halt("DistributedBag Internal Error: 'elems' is nil");
+    }
+
+    if isFull {
+      halt("DistributedBag Internal Error: SegmentBlock is Full");
+    }
+
+    elems[size] = elt;
+    size = size + 1;
+  }
+
+  inline proc pop() : eltType {
+    if elems == nil {
+      halt("DistributedBag Internal Error: 'elems' is nil");
+    }
+
+    if isEmpty {
+      halt("DistributedBag Internal Error: SegmentBlock is Empty");
+    }
+
+    size = size - 1;
+    var elt = elems[size];
+    return elt;
+  }
+
+  proc BagSegmentBlock(type eltType, capacity) {
+    if capacity == 0 {
+      halt("DistributedBag Internal Error: Capacity is 0...");
+    }
+
+    cap = capacity;
+    elems = c_malloc(eltType, capacity);

```

Comment:
A few things here. Originally, I used Chapel Arrays and found that after making the switch to using tuples, I found a significant performance improvement (I believe Chapel perfs show the comparisong [here](http://chapel.sourceforge.net/perf/chapcs/?startdate=2015/09/10&enddate=2017/08/23&graphs=arrayvstupleserialaccesses)) after the switch (at least 4 - 5x improvement). After deciding to double the size of each successor unroll block, I saw an improvement of at least 1.5x by going from Tuples to C_Ptr. Those improvements weren't even factoring in bulk transfer, that was just from adding elements one at a time.

We also use bulk transfer across nodes for load balancing (dynamic load balancing). Currently, the explicit transfers definitely could be improved a bit for more performance, like implementing some direct zero-copy from one segment to another, but overall C_Ptrs have dramatically improved performance. Even if array bulk transfers do not cause a significant performance degradation, I can see it causing issues in other areas. Especially while adding a large number of elements, and even during load balancing where we likely would create a significant number of unroll blocks.